resteasy-client-microprofile-base/src/main/java/org/jboss/resteasy/microprofile/client/RestClientBuilderImpl.java

```diff
@@ -242,49 +253,27 @@ public RestClientBuilder executorService(ExecutorService executor) {
-            boolean isUriMatched = false;
-            if (envProxyHost != null && !noProxyHosts.isEmpty()) {
-                for (String s : noProxyHosts) {
-                    Pattern p = Pattern.compile(s);
-                    Matcher m = p.matcher(baseURI.getHost());
-                    isUriMatched = m.matches();
-                    if (isUriMatched) {
-                        break;
-                    }
-                }
-            }
-
-            if (envProxyHost != null && !isUriMatched) {
-                // Use proxy, if defined in the env variables
-                resteasyClientBuilder = builderDelegate.defaultProxy(envProxyHost,
-                        Integer.parseInt(getSystemProperty("http.proxyPort", "80")));
-            } else {
-                // Search for proxy settings passed in the client builder, if passed and use them if found
-                String userProxyHost = Optional.ofNullable(getConfiguration().getProperty(PROPERTY_PROXY_HOST))
-                        .filter(String.class::isInstance).map(String.class::cast).orElse(null);
+            String userProxyHost = Optional.ofNullable(getConfiguration().getProperty(PROPERTY_PROXY_HOST))
+                    .filter(String.class::isInstance).map(String.class::cast).orElse(null);
 
-                Integer userProxyPort = Optional.ofNullable(getConfiguration().getProperty(PROPERTY_PROXY_PORT))
-                        .filter(Integer.class::isInstance).map(Integer.class::cast).orElse(null);
+            Integer userProxyPort = Optional.ofNullable(getConfiguration().getProperty(PROPERTY_PROXY_PORT))
+                    .filter(Integer.class::isInstance).map(Integer.class::cast).orElse(null);
 
+            // configuration of the builder has precedence over environment variables
+            if (userProxyHost != null && userProxyPort != null) {
                 String userProxyScheme = Optional.ofNullable(getConfiguration().getProperty(PROPERTY_PROXY_SCHEME))
                         .filter(String.class::isInstance).map(String.class::cast).orElse(null);
 
-                if (userProxyHost != null && userProxyPort != null) {
-                    resteasyClientBuilder = builderDelegate.defaultProxy(userProxyHost, userProxyPort, userProxyScheme);
-                } else {
-                    // ProxySelector if applicable
-                    selectHttpProxy().ifPresent(
-                            proxyAddress -> builderDelegate.defaultProxy(proxyAddress.getHostString(), proxyAddress.getPort()));
-
-                    resteasyClientBuilder = builderDelegate;
-                }
+                resteasyClientBuilder.defaultProxy(userProxyHost, userProxyPort, userProxyScheme);
+            } else {
+                selectHttpProxy().ifPresent(
+                        proxyAddress -> resteasyClientBuilder.defaultProxy(proxyAddress.getHostString(),
+                                proxyAddress.getPort()));
```

Comment:
For that change, better have a look at the result as it's far more simple than the original code.
You can see a reference to the issues it fixes here: https://github.com/quarkusio/quarkus/pull/19261 . The problem is mostly that you consider that `http.nonProxyHosts` contains regexps, which is incorrect. Your default config of `localhost|127.*|[::1]` is by miracle valid regexps but it's not really testing things properly.

With this patch we delegate everything to the `ProxySelector` which is doing the right thing.

Whatever we decide for short term I think we need this patch in RESTEasy long term.